base/util/PKICertImport.bash

```diff
@@ -0,0 +1,490 @@
+        # to keep secure when copying. This ensures other users cannot access
+        # our copied files.
+
+        local user=""
+        local group=""
+
+        # Acquire curreent username.
+        user="$(id --user --name 2>&1)"
+        ret=$?
+        if (( ret != 0 )); then
+            __e "id exited with non-zero result: $ret" 1>&2
+            __e "Unable to get current user's name." 1>&2
+            __rm_secure_tmp
+            exit 1
+        fi
+
+        # Acquire curreent primary group.
+        group="$(id --group --name 2>&1)"
+        ret=$?
+        if (( ret != 0 )); then
+            __e "id exited with non-zero result: $ret" 1>&2
+            __e "Unable to get current user's name." 1>&2
+            __rm_secure_tmp
+            exit 1
+        fi
+
+        # Change ownership prior to permissions; theoretically these should
+        # already be the current owner.
+        chown "$user:$group" -R "$TMPBASE"
+        ret=$?
+        if (( ret != 0 )); then
+            __e "Return from chown on $TMPBASE was non-zero: $ret" 1>&2
+            __rm_secure_tmp
+            exit 1
+        fi
+
+        # Restrict access only to the owner, preventing any group and world
+        # access.
+        chmod 700 -R "$TMPBASE"
+        ret=$?
+        if (( ret != 0 )); then
+            __e "Return from chmod on $TMPBASE was non-zero: $ret" 1>&2
+            __rm_secure_tmp
+            exit 1
+        fi
+
+        return 0
+    }
+
+    ## rm_secure_tmp removes the temporary directory if present.
+    function __rm_secure_tmp() {
+        if [ -d "$TMPBASE" ]; then
+            # TODO: Determine if we should shred the contents if the shred
+            # utility is present; otherwise, could fall back on dd...
+            rm -rf "$TMPBASE"
+        fi
+    }
+
+
+    ## [ core commands ] ##
+
+    # Parse the command line arguments and set the appropriate global
+    # variables. Return status of non-zero indicates a fatal error; help
+    # should be displayed. Return status of zero indicates no error and help
+    # should not be displayed.
+    function _parse_args() {
+        # Use a read-and-shift approach to parse both "<option>" and
+        # "<option> <value>" forms.
+        while (( $# > 0 )); do
+            local arg="$1"
+            shift
+
+            # Sorted alphabetically by short option.
+            if [ "x$arg" == "x--ascii" ] || [ "x$arg" == "x-a" ]; then
+                # If specified, the -a flag is passed when the certificate is
+                # imported.
+                CERT_ASCII="true"
+            elif [ "x$arg" == "x--database" ] || [ "x$arg" == "x-d" ]; then
+                # Always required; path to the original NSS DB. Note that this
+                # differs from certutil in that we detect the NSSDB type here,
+                # versus taking a prefix:path combination.
+                NSSDB="$1"
+
+                if [ -e "$NSSDB/cert8.db" ] && [ ! -e "$NSSDB/cert9.db" ]; then
```

Comment:
@ladycfu This detection code is necessary on older `certutil` versions. In particular, if you've setup a NSS DB with SQL (`certutil -N -d sql:~/nssdb`) and then perform an add command without specifying the type (`certutil -A -d ~/nssdb`), it'll go back and create a `dbm`-style nssdb in the same directory and cause problems for itself since the two (`sql` and `dbm` style nssdbs) aren't synced after that... Yucky!

Reply:
yes I saw.  looks fine.